// @aws-cdk/aws-route53/test/test.hosted-zone-provider.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/core");
const lib_1 = require("../lib");
module.exports = {
    'Hosted Zone Provider': {
        'HostedZoneProvider will return context values if available'(test) {
            // GIVEN
            const stack = new cdk.Stack(undefined, 'TestStack', {
                env: { account: '12345', region: 'us-east-1' }
            });
            const filter = { domainName: 'test.com' };
            lib_1.HostedZone.fromLookup(stack, 'Ref', filter);
            const missing = assert_1.SynthUtils.synthesize(stack).assembly.manifest.missing;
            test.ok(missing && missing.length === 1);
            const fakeZoneId = '11111111111111';
            const fakeZone = {
                Id: `/hostedzone/${fakeZoneId}`,
                Name: 'example.com.',
                CallerReference: 'TestLates-PublicZo-OESZPDFV7G6A',
                Config: {
                    Comment: 'CDK created',
                    PrivateZone: false
                },
                ResourceRecordSetCount: 3
            };
            const stack2 = new cdk.Stack(undefined, 'TestStack', {
                env: { account: '12345', region: 'us-east-1' }
            });
            stack2.node.setContext(missing[0].key, fakeZone);
            // WHEN
            const zoneRef = lib_1.HostedZone.fromLookup(stack2, 'MyZoneProvider', filter);
            // THEN
            test.deepEqual(zoneRef.hostedZoneId, fakeZoneId);
            test.done();
        },
        'HostedZoneProvider will return context values if available when using plain hosted zone id'(test) {
            // GIVEN
            const stack = new cdk.Stack(undefined, 'TestStack', {
                env: { account: '12345', region: 'us-east-1' }
            });
            const filter = { domainName: 'test.com' };
            lib_1.HostedZone.fromLookup(stack, 'Ref', filter);
            const missing = assert_1.SynthUtils.synthesize(stack).assembly.manifest.missing;
            test.ok(missing && missing.length === 1);
            const fakeZoneId = '11111111111111';
            const fakeZone = {
                Id: `/hostedzone/${fakeZoneId}`,
                Name: 'example.com.',
                CallerReference: 'TestLates-PublicZo-OESZPDFV7G6A',
                Config: {
                    Comment: 'CDK created',
                    PrivateZone: false
                },
                ResourceRecordSetCount: 3
            };
            const stack2 = new cdk.Stack(undefined, 'TestStack', {
                env: { account: '12345', region: 'us-east-1' }
            });
            stack2.node.setContext(missing[0].key, fakeZone);
            const zone = lib_1.HostedZone.fromLookup(stack2, 'MyZoneProvider', filter);
            // WHEN
            const zoneId = zone.hostedZoneId;
            // THEN
            test.deepEqual(fakeZoneId, zoneId);
            test.done();
        }
    }
};
