// @aws-cdk/aws-kms/test/test.via-service-principal.js
"use strict";
const iam = require("@aws-cdk/aws-iam");
const kms = require("../lib");
module.exports = {
    'Via service, any principal'(test) {
        // WHEN
        const statement = new iam.PolicyStatement({
            actions: ['abc:call'],
            principals: [new kms.ViaServicePrincipal('bla.amazonaws.com')],
            resources: ['*']
        });
        // THEN
        test.deepEqual(statement.toStatementJson(), {
            Action: 'abc:call',
            Condition: { StringEquals: { 'kms:ViaService': 'bla.amazonaws.com' } },
            Effect: 'Allow',
            Principal: '*',
            Resource: '*'
        });
        test.done();
    },
    'Via service, principal with conditions'(test) {
        // WHEN
        const statement = new iam.PolicyStatement({
            actions: ['abc:call'],
            principals: [new kms.ViaServicePrincipal('bla.amazonaws.com', new iam.OrganizationPrincipal('o-1234'))],
            resources: ['*']
        });
        // THEN
        test.deepEqual(statement.toStatementJson(), {
            Action: 'abc:call',
            Condition: {
                StringEquals: {
                    'kms:ViaService': 'bla.amazonaws.com',
                    'aws:PrincipalOrgID': 'o-1234'
                }
            },
            Effect: 'Allow',
            Principal: '*',
            Resource: '*'
        });
        test.done();
    },
};
