// @aws-cdk/assets/test/fs/test.utils.js
"use strict";
const fs = require("fs");
const path = require("path");
const ts_mock_imports_1 = require("ts-mock-imports");
const fs_1 = require("../../lib/fs");
const util = require("../../lib/fs/utils");
module.exports = {
    shouldExclude: {
        'excludes nothing by default'(test) {
            test.ok(!util.shouldExclude([], path.join('some', 'file', 'path')));
            test.done();
        },
        'excludes requested files'(test) {
            const exclusions = ['*.ignored'];
            test.ok(util.shouldExclude(exclusions, path.join('some', 'file.ignored')));
            test.ok(!util.shouldExclude(exclusions, path.join('some', 'important', 'file')));
            test.done();
        },
        'does not exclude whitelisted files'(test) {
            const exclusions = ['*.ignored', '!important.*'];
            test.ok(!util.shouldExclude(exclusions, path.join('some', 'important.ignored')));
            test.done();
        },
    },
    shouldFollow: {
        always: {
            'follows internal'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join(sourceRoot, 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync', true);
                try {
                    test.ok(util.shouldFollow(fs_1.FollowMode.ALWAYS, sourceRoot, linkTarget));
                    test.ok(mockFsExists.calledOnceWith(linkTarget));
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
            'follows external'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join('alternate', 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync', true);
                try {
                    test.ok(util.shouldFollow(fs_1.FollowMode.ALWAYS, sourceRoot, linkTarget));
                    test.ok(mockFsExists.calledOnceWith(linkTarget));
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
            'does not follow internal when the referent does not exist'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join(sourceRoot, 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync', false);
                try {
                    test.ok(!util.shouldFollow(fs_1.FollowMode.ALWAYS, sourceRoot, linkTarget));
                    test.ok(mockFsExists.calledOnceWith(linkTarget));
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
            'does not follow external when the referent does not exist'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join('alternate', 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync', false);
                try {
                    test.ok(!util.shouldFollow(fs_1.FollowMode.ALWAYS, sourceRoot, linkTarget));
                    test.ok(mockFsExists.calledOnceWith(linkTarget));
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
        },
        external: {
            'does not follow internal'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join(sourceRoot, 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync');
                try {
                    test.ok(!util.shouldFollow(fs_1.FollowMode.EXTERNAL, sourceRoot, linkTarget));
                    test.ok(mockFsExists.notCalled);
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
            'follows external'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join('alternate', 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync', true);
                try {
                    test.ok(util.shouldFollow(fs_1.FollowMode.EXTERNAL, sourceRoot, linkTarget));
                    test.ok(mockFsExists.calledOnceWith(linkTarget));
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
            'does not follow external when referent does not exist'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join('alternate', 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync', false);
                try {
                    test.ok(!util.shouldFollow(fs_1.FollowMode.EXTERNAL, sourceRoot, linkTarget));
                    test.ok(mockFsExists.calledOnceWith(linkTarget));
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
        },
        blockExternal: {
            'follows internal'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join(sourceRoot, 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync', true);
                try {
                    test.ok(util.shouldFollow(fs_1.FollowMode.BLOCK_EXTERNAL, sourceRoot, linkTarget));
                    test.ok(mockFsExists.calledOnceWith(linkTarget));
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
            'does not follow internal when referent does not exist'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join(sourceRoot, 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync', false);
                try {
                    test.ok(!util.shouldFollow(fs_1.FollowMode.BLOCK_EXTERNAL, sourceRoot, linkTarget));
                    test.ok(mockFsExists.calledOnceWith(linkTarget));
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
            'does not follow external'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join('alternate', 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync');
                try {
                    test.ok(!util.shouldFollow(fs_1.FollowMode.BLOCK_EXTERNAL, sourceRoot, linkTarget));
                    test.ok(mockFsExists.notCalled);
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
        },
        never: {
            'does not follow internal'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join(sourceRoot, 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync');
                try {
                    test.ok(!util.shouldFollow(fs_1.FollowMode.NEVER, sourceRoot, linkTarget));
                    test.ok(mockFsExists.notCalled);
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
            'does not follow external'(test) {
                const sourceRoot = path.join('source', 'root');
                const linkTarget = path.join('alternate', 'referent');
                const mockFsExists = ts_mock_imports_1.ImportMock.mockFunction(fs, 'existsSync');
                try {
                    test.ok(!util.shouldFollow(fs_1.FollowMode.NEVER, sourceRoot, linkTarget));
                    test.ok(mockFsExists.notCalled);
                    test.done();
                }
                finally {
                    mockFsExists.restore();
                }
            },
        }
    },
};
