// @aws-cdk/aws-iam/test/user.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@aws-cdk/assert/jest");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
describe('IAM user', () => {
    test('default user', () => {
        const app = new core_1.App();
        const stack = new core_1.Stack(app, 'MyStack');
        new lib_1.User(stack, 'MyUser');
        expect(stack).toMatchTemplate({
            Resources: { MyUserDC45028B: { Type: 'AWS::IAM::User' } }
        });
    });
    test('default user with password', () => {
        const app = new core_1.App();
        const stack = new core_1.Stack(app, 'MyStack');
        new lib_1.User(stack, 'MyUser', {
            password: core_1.SecretValue.plainText('1234')
        });
        expect(stack).toMatchTemplate({
            Resources: {
                MyUserDC45028B: {
                    Type: 'AWS::IAM::User',
                    Properties: { LoginProfile: { Password: '1234' } }
                }
            }
        });
    });
    test('fails if reset password is required but no password is set', () => {
        const app = new core_1.App();
        const stack = new core_1.Stack(app, 'MyStack');
        expect(() => new lib_1.User(stack, 'MyUser', { passwordResetRequired: true })).toThrow();
    });
    test('create with managed policy', () => {
        // GIVEN
        const app = new core_1.App();
        const stack = new core_1.Stack(app, 'MyStack');
        // WHEN
        new lib_1.User(stack, 'MyUser', {
            managedPolicies: [lib_1.ManagedPolicy.fromAwsManagedPolicyName('asdf')]
        });
        // THEN
        expect(stack).toHaveResource('AWS::IAM::User', {
            ManagedPolicyArns: [
                { "Fn::Join": ["", ["arn:", { Ref: "AWS::Partition" }, ":iam::aws:policy/asdf"]] }
            ]
        });
    });
    test('can supply permissions boundary managed policy', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const permissionsBoundary = lib_1.ManagedPolicy.fromAwsManagedPolicyName('managed-policy');
        new lib_1.User(stack, 'MyUser', {
            permissionsBoundary,
        });
        expect(stack).toHaveResource('AWS::IAM::User', {
            PermissionsBoundary: {
                "Fn::Join": [
                    "",
                    [
                        "arn:",
                        {
                            Ref: "AWS::Partition"
                        },
                        ":iam::aws:policy/managed-policy"
                    ]
                ]
            }
        });
    });
    test('imported user has an ARN', () => {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        const user = lib_1.User.fromUserName(stack, 'import', 'MyUserName');
        // THEN
        expect(stack.resolve(user.userArn)).toStrictEqual({
            "Fn::Join": ["", ["arn:", { Ref: "AWS::Partition" }, ":iam::", { Ref: "AWS::AccountId" }, ":user/MyUserName"]]
        });
    });
});
