// @aws-cdk/aws-events/test/test.util.js
"use strict";
const util_1 = require("../lib/util");
module.exports = {
    mergeEventPattern: {
        'happy case'(test) {
            test.deepEqual(util_1.mergeEventPattern({
                bar: [1, 2],
                hey: ['happy'],
                hello: {
                    world: ['hi', 'dude'],
                    case: [1]
                }
            }, {
                hey: ['day', 'today'],
                hello: {
                    world: ['you']
                }
            }), {
                bar: [1, 2],
                hey: ['happy', 'day', 'today'],
                hello: {
                    world: ['hi', 'dude', 'you'],
                    case: [1]
                }
            });
            test.done();
        },
        'merge into an empty destination'(test) {
            test.deepEqual(util_1.mergeEventPattern(undefined, { foo: ['123'] }), { foo: [123] });
            test.deepEqual(util_1.mergeEventPattern(undefined, { foo: { bar: ['123'] } }), { foo: { bar: [123] } });
            test.deepEqual(util_1.mergeEventPattern({}, { foo: { bar: ['123'] } }), { foo: { bar: [123] } });
            test.done();
        },
        'fails if a field is not an array'(test) {
            test.throws(() => util_1.mergeEventPattern(undefined, 123), /Invalid event pattern '123', expecting an object or an array/);
            test.throws(() => util_1.mergeEventPattern(undefined, 'Hello'), /Invalid event pattern '"Hello"', expecting an object or an array/);
            test.throws(() => util_1.mergeEventPattern(undefined, { foo: '123' }), /Invalid event pattern field { foo: "123" }. All fields must be arrays/);
            test.done();
        },
        'fails if mismatch between dest and src'(test) {
            test.throws(() => util_1.mergeEventPattern({
                obj: {
                    array: [1]
                }
            }, {
                obj: {
                    array: {
                        value: ['hello']
                    }
                }
            }), /Invalid event pattern field array. Type mismatch between existing pattern \[1\] and added pattern \{"value":\["hello"\]\}/);
            test.done();
        }
    }
};
