// @aws-cdk/aws-s3/test/test.metrics.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
module.exports = {
    'Can use addMetrics() to add a metric configuration'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        const bucket = new lib_1.Bucket(stack, 'Bucket');
        bucket.addMetric({
            id: "test"
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::S3::Bucket', {
            MetricsConfigurations: [{
                    Id: "test"
                }]
        }));
        test.done();
    },
    'Bucket with metrics on prefix'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.Bucket(stack, 'Bucket', {
            metrics: [{
                    id: "test",
                    prefix: "prefix"
                }]
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::S3::Bucket', {
            MetricsConfigurations: [{
                    Id: "test",
                    Prefix: "prefix"
                }]
        }));
        test.done();
    },
    'Bucket with metrics on tag filter'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.Bucket(stack, 'Bucket', {
            metrics: [{
                    id: "test",
                    tagFilters: { tagname1: "tagvalue1", tagname2: "tagvalue2" }
                }]
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::S3::Bucket', {
            MetricsConfigurations: [{
                    Id: "test",
                    TagFilters: [
                        { Key: "tagname1", Value: "tagvalue1" },
                        { Key: "tagname2", Value: "tagvalue2" },
                    ]
                }]
        }));
        test.done();
    },
    'Bucket with multiple metric configurations'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.Bucket(stack, 'Bucket', {
            metrics: [
                {
                    id: "test",
                    tagFilters: { tagname1: "tagvalue1", tagname2: "tagvalue2" }
                },
                {
                    id: "test2",
                    prefix: "prefix"
                },
            ]
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::S3::Bucket', {
            MetricsConfigurations: [{
                    Id: "test",
                    TagFilters: [
                        { Key: "tagname1", Value: "tagvalue1" },
                        { Key: "tagname2", Value: "tagvalue2" },
                    ]
                },
                {
                    Id: "test2",
                    Prefix: "prefix"
                }]
        }));
        test.done();
    },
};
