// @aws-cdk/aws-s3/test/test.cors.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
module.exports = {
    'Can use addCors() to add a CORS configuration'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        const bucket = new lib_1.Bucket(stack, 'Bucket');
        bucket.addCorsRule({
            allowedMethods: [lib_1.HttpMethods.GET, lib_1.HttpMethods.HEAD],
            allowedOrigins: ["https://example.com"]
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::S3::Bucket', {
            CorsConfiguration: {
                CorsRules: [{
                        AllowedMethods: ["GET", "HEAD"],
                        AllowedOrigins: ["https://example.com"]
                    }]
            }
        }));
        test.done();
    },
    'Bucket with multiple cors configurations'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.Bucket(stack, 'Bucket', {
            cors: [
                {
                    allowedHeaders: [
                        "*"
                    ],
                    allowedMethods: [
                        lib_1.HttpMethods.GET
                    ],
                    allowedOrigins: [
                        "*"
                    ],
                    exposedHeaders: [
                        "Date"
                    ],
                    id: "myCORSRuleId1",
                    maxAge: 3600
                },
                {
                    allowedHeaders: [
                        "x-amz-*"
                    ],
                    allowedMethods: [
                        lib_1.HttpMethods.DELETE
                    ],
                    allowedOrigins: [
                        "http://www.example1.com",
                        "http://www.example2.com"
                    ],
                    exposedHeaders: [
                        "Connection",
                        "Server",
                        "Date"
                    ],
                    id: "myCORSRuleId2",
                    maxAge: 1800
                }
            ]
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::S3::Bucket', {
            CorsConfiguration: {
                CorsRules: [
                    {
                        AllowedHeaders: [
                            "*"
                        ],
                        AllowedMethods: [
                            "GET"
                        ],
                        AllowedOrigins: [
                            "*"
                        ],
                        ExposedHeaders: [
                            "Date"
                        ],
                        Id: "myCORSRuleId1",
                        MaxAge: 3600
                    },
                    {
                        AllowedHeaders: [
                            "x-amz-*"
                        ],
                        AllowedMethods: [
                            "DELETE"
                        ],
                        AllowedOrigins: [
                            "http://www.example1.com",
                            "http://www.example2.com"
                        ],
                        ExposedHeaders: [
                            "Connection",
                            "Server",
                            "Date"
                        ],
                        Id: "myCORSRuleId2",
                        MaxAge: 1800
                    }
                ]
            }
        }));
        test.done();
    },
};
