// @aws-cdk/aws-cloudformation/test/example.customresource.lit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lambda = require("@aws-cdk/aws-lambda");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
class CopyOperation extends core_1.Construct {
    constructor(parent, name, props) {
        super(parent, name);
        const lambdaProvider = new lambda.SingletonFunction(this, 'Provider', {
            uuid: 'f7d4f730-4ee1-11e8-9c2d-fa7ae01bbebc',
            runtime: lambda.Runtime.PYTHON_3_7,
            code: lambda.Code.fromAsset('../copy-handler'),
            handler: 'index.handler',
            timeout: core_1.Duration.seconds(60),
        });
        new lib_1.CustomResource(this, 'Resource', {
            provider: lib_1.CustomResourceProvider.fromLambda(lambdaProvider),
            properties: {
                sourceBucketArn: props.sourceBucket.bucketArn,
                targetBucketArn: props.targetBucket.bucketArn,
            }
        });
    }
}
/// !hide
new CopyOperation(new core_1.Stack(), 'Test', {
    sourceBucket: { bucketArn: 'x' },
    targetBucket: { bucketArn: 'y' },
});
