// @aws-cdk/cx-api/lib/metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Metadata key used to print INFO-level messages by the toolkit when an app is syntheized.
 */
exports.INFO_METADATA_KEY = 'aws:cdk:info';
/**
 * Metadata key used to print WARNING-level messages by the toolkit when an app is syntheized.
 */
exports.WARNING_METADATA_KEY = 'aws:cdk:warning';
/**
 * Metadata key used to print ERROR-level messages by the toolkit when an app is syntheized.
 */
exports.ERROR_METADATA_KEY = 'aws:cdk:error';
/**
 * The key used when CDK path is embedded in **CloudFormation template** metadata (not cdk metadata).
 */
exports.PATH_METADATA_KEY = 'aws:cdk:path';
/**
 * Represents the CloudFormation logical ID of a resource at a certain path.
 */
exports.LOGICAL_ID_METADATA_KEY = 'aws:cdk:logicalId';
/**
 * Tag metadata key.
 */
exports.STACK_TAGS_METADATA_KEY = 'aws:cdk:stack-tags';
var SynthesisMessageLevel;
(function (SynthesisMessageLevel) {
    SynthesisMessageLevel["INFO"] = "info";
    SynthesisMessageLevel["WARNING"] = "warning";
    SynthesisMessageLevel["ERROR"] = "error";
})(SynthesisMessageLevel = exports.SynthesisMessageLevel || (exports.SynthesisMessageLevel = {}));
