// @aws-cdk/aws-apigateway/test/integ.restapi.multiuse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lambda = require("@aws-cdk/aws-lambda");
const cdk = require("@aws-cdk/core");
const apigw = require("../lib");
class MultiStack extends cdk.Stack {
    constructor(scope, id) {
        super(scope, id);
        const hello = new apigw.LambdaIntegration(new lambda.Function(this, 'Hello', {
            runtime: lambda.Runtime.NODEJS_10_X,
            handler: 'index.handler',
            code: lambda.Code.inline(`exports.handler = ${helloCode}`)
        }));
        const api = new apigw.RestApi(this, 'hello-api');
        api.root.resourceForPath('/hello').addMethod('GET', hello);
        const api2 = new apigw.RestApi(this, 'second-api');
        api2.root.resourceForPath('/hello').addMethod('GET', hello);
    }
}
class MultiApp extends cdk.App {
    constructor() {
        super();
        new MultiStack(this, 'restapi-multiuse-example');
    }
}
function helloCode(_event, _context, callback) {
    return callback(undefined, {
        statusCode: 200,
        body: 'hello, world!'
    });
}
new MultiApp().synth();
