// @aws-cdk/aws-cloudwatch/test/test.metrics.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const iam = require("@aws-cdk/aws-iam");
const cdk = require("@aws-cdk/core");
const lib_1 = require("../lib");
module.exports = {
    'metric grant'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const role = new iam.Role(stack, 'SomeRole', {
            assumedBy: new iam.Anyone()
        });
        // WHEN
        lib_1.Metric.grantPutMetricData(role);
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Policy', {
            PolicyDocument: {
                Version: "2012-10-17",
                Statement: [
                    {
                        Action: "cloudwatch:PutMetricData",
                        Effect: "Allow",
                        Resource: "*"
                    }
                ],
            },
        }));
        test.done();
    },
    'can not use invalid period in Metric'(test) {
        test.throws(() => {
            new lib_1.Metric({ namespace: 'Test', metricName: 'ACount', period: cdk.Duration.seconds(20) });
        }, /'period' must be 1, 5, 10, 30, or a multiple of 60 seconds, received 20/);
        test.done();
    },
    'Metric optimization: "with" with the same period returns the same object'(test) {
        const m = new lib_1.Metric({ namespace: 'Test', metricName: 'Metric', period: cdk.Duration.minutes(10) });
        // Note: object equality, NOT deep equality on purpose
        test.equals(m.with({}), m);
        test.equals(m.with({ period: cdk.Duration.minutes(10) }), m);
        test.notEqual(m.with({ period: cdk.Duration.minutes(5) }), m);
        test.done();
    },
};
